packages/fdc3-agent-proxy/src/intents/DefaultIntentSupport.ts

```diff
@@ -163,26 +169,38 @@ export class DefaultIntentSupport implements IntentSupport {
       meta: this.messaging.createMeta(),
     };
 
-    const resultPromise = this.createResultPromise(messageOut);
-    const response: RaiseIntentForContextResponse = await this.messaging.exchange(
-      messageOut,
+    const resultPromise = this.createResultPromise(request);
+    const response = await this.messaging.exchange<RaiseIntentForContextResponse>(
+      request,
       'raiseIntentForContextResponse',
       ResolveError.IntentDeliveryFailed
     );
+
+    throwIfUndefined(
+      response.payload.error ?? response.payload.appIntents ?? response.payload.intentResolution,
+      'Invalid response from Desktop Agent to raiseIntentForContext!',
+      response,
+      ResolveError.NoAppsFound
+    );
+
     if (response.payload.appIntents) {
-      // we need to invoke the resolver
+      // Needs further resolution, we need to invoke the resolver
       const result: IntentResolutionChoice | void = await this.intentResolver.chooseIntent(
-        response.payload.appIntents as any[],
+        response.payload.appIntents,
         context
       );
       if (result) {
         return this.raiseIntent(result.intent, context, result.appId);
       } else {
         throw new Error(ResolveError.UserCancelled);
       }
-    } else {
-      const details = response.payload.intentResolution!;
+    } else if (response.payload.intentResolution) {
+      // Was resolved
+      const details = response.payload.intentResolution;
       return new DefaultIntentResolution(this.messaging, resultPromise, details.source, details.intent);
+    } else {
+      //should never get here as we will throw above
+      throw new Error(response.payload.error);
```

Comment:
we can never reach this line, there's something fishy about this

Reply:
`throwIfUndefined` falls down a bit when we have more than one property.  In this case (and the one above) wouldn't the answer be for this not to be an `else if` but just an `if` and no `else`?  


i.e. logically we know either intentResolution or appIntents is set (because of the `throwIfUndefined`, so just have an if with out else for each.

Reply:
If we've got to validate two properties then you just put two `throwIfUndefined` calls. In this case the properties are mutually exclusive so coalescing them in a single test works. We can drop the final else (still needs an else if due to the two different properties).

There's no perf penalty to having the else and it remains defensive (if someone breaks exchange), but we might as well focus on making sure that doesn't happen and go for brevity. Will cut the else.

Reply:
P.S. added to `exchange` and `waitFor`'s jsDoc to state that they will reject with errors in payloads.

Reply:
Ah... Without the else clause typescript can't confirm that all branches return (so it can't be if/elseif ir if/if, it has to be if/else if/else. Putting 'er back.

Reply:
Ok, let me explain why this is relevant and not cost-free.  As we are measuring code coverage, any time we have code that is less than 100% covered, it adds an overhead to the reviewers (i.e. us).  We have to review the files and see why code coverage isn't complete, and then decide manually whether to accept or reject it.   

My experience is that this adds significant overhead to the review process especially when you have large codebases and regular PRs coming in.  

You can apply this thinking to a lot of my suggestions above.  So, you have to weigh what is going to save us time in the long run - extra effort for reviewing PRs or possibly extra effort debugging.

Reply:
I don't think this is a weakness in typescript, if you've got a branch that doesn't return (implictly returning void) and violates the return type specified that is a legitimate problem - that code wouldn't compile in most languages. The `throwIfUndefined` function we've added (to increase coverage) is part of the issue (in these cases with multiple properties / differnet returns) as typescript is not seeing the test and thrown error it contains, which could eliminate the branch. However, we're in a better state with it. 

Using a type override or similar to eliminate the coverage issue (how it was previously) is probably worse as you've got to parse the code yourself to spot a potential issue, with coverage perhaps providing a false sense of security.

That leaves... an annotation to ignore these else clauses? I've installed one on each one flagged in these reviews. Note they have to go before the if statement to catch the whole else:

```ts
/* istanbul ignore else */
if (response.payload.appIntents) {
  ...
} else if (...) {
  ...
} else {    //ignored
  ...       //ignored
}           //ignored
```

Such things should be used _very_ sparingly!

Reply:
> I don't think this is a weakness in typescript,  

So we know from this line that either appIntents or intentResolution is set:

```
throwIfUndefined(
      response.payload.error ?? response.payload.appIntent ?? response.payload.intentResolution,
      'Invalid response from Desktop Agent to raiseIntent!',
      response,
      ResolveError.NoAppsFound
    )
```

Therefore if this isn't true:

```
 if (response.payload.appIntent) {
      // Needs further resolution, we need to invoke the resolver
  ...
```

Then intentResolution must be set, so you can do:

```
 } else {
      // Was resolved
      const details = response.payload.intentResolution!;
      return new DefaultIntentResolution(this.messaging, resultPromise, details.source, details.intent);
    }
```

There are good uses for `!` - the typescript type system can't do everything, including this case where one of two things is definitely set. It just models that as both are optional.

Reply:
Fair enough, given that the original issue is solved either way (validating the DA's response) we can go with whats shorter!